Fixes #28099 - Adding total_disk_size to host facets

domitea opened the pull request:
This PR introduced new parameter to hosts facets. Parameter is a total disk size that is included in Discovery plugin at own. However that means that we have to maintain a two "facets like" codes at one time (one at core, another one at discovery). This is attempt to unify that and having only one code for that.


<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Mark all strings for translation, see [1]
* Suggest prerequisites for testing and testing scenarios following example above.
* Prepend `[WIP]` for work in progress to prevent bots from triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* List all prerequisites for testing (e.g. VMware cluster, two smart proxies...)
* Reviewers often use extensive list of items to check, have a look prior submitting [2]
* Be nice and respectful

1: https://projects.theforeman.org/projects/foreman/wiki/Translating#Translating-for-developers
2: https://github.com/theforeman/foreman/blob/develop/developer_docs/pr_review.asciidoc
-->

domitea committed:
Fixes #32170 - Adding total_disk_size to host facets

This PR introduced new parameter to hosts facets. Parameter is a total
disk size that is included in Discovery plugin at own. However that
means that we have to maintain a two "facets like" codes at one time
(one at core, another one at discovery). This is attempt to unify
that and having only one code for that.

lzap commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,17 @@ def cores
     facts.dig('processors', 'count') || facts['processorcount']
   end
 
+  def total_disk_size
```
To have similar naming as other facets, I suggest to use `disks_total`.

lzap commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,17 @@ def cores
 
+  def total_disk_size
+    disks      = facts.select { |key, _| key.to_s =~ /blockdevice.*_size/ }
+    disks_size = 0
```
This method must work also with structured facts, because discovery only supports Facter 2.X but core supports Facter 3.x and above. See above for an example:

```
  def ram
    if (value = facts.dig('memory', 'system', 'total_bytes'))
      value / 1.megabyte
    else
      facts['memorysize_mb']
    end
  end
```

ares replied:
The discovery now uses facter 4 right?

lzap reviewed (commented):
Please add a tests - both for facter 2.x and 3.x, we should have some JSON fixture in the tests folder.

lzap commented:
Hmm our fact fixtures are in `test/static_fixtures/facts/` but it looks like we don't have any legacy versions there. If you install "facter" rubygem and call `facter --json > example.json` for the following versions: 

* 2.0.2
* 2.5.7
* 3.9.6
* 3.12.2
* 4.0.52

It would be great to have fixtures for a single system but with all these different facter versions. Then you can write a test for all versions making sure disks_total works fine for all facter versions. Currently we only do this randomly in `test/unit/puppet_fact_parser_test.rb` - some methods are tested against RHEL v2, others against Debian v2 or v3. It would be huge help if you could rewrite this test case to work with all mentioned versions (e.g. in a loop for every single one).

https://rubygems.org/gems/facter/versions

ares commented:
Also please clone the ticket from the discovery project to core and use that, we need 2 redmine issues for patches in 2 repos.

ares commented:
Since we now introduce this as a new core fact, we should also take a look and implementing this for other facters
* katello https://github.com/Katello/katello/blob/master/app/models/katello/rhsm_fact_parser.rb#L96-L106
* ansible https://github.com/theforeman/foreman_ansible/blob/master/app/services/foreman_ansible/fact_parser.rb - we seem to be missing other core facts implementation there :-(
* chef https://github.com/theforeman/foreman_chef/blob/master/app/models/foreman_chef/fact_parser.rb#L95-L97 (the same as for ansible)
* salt https://github.com/theforeman/foreman_salt/blob/78e78ea981bc173a7280ada6db477e0ad79ce108/app/services/foreman_salt/fact_parser.rb - this does not seem to implement any

Opening issue for respective projects and linking them here may be good enough.

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,17 @@ def cores
   end
 
+  def total_disk_size
+    disks      = facts.select { |key, _| key.to_s =~ /blockdevice.*_size/ }
```
I think this is confusing. The method is named `total_disk_size` but here you look at partitions. It can also be confusing if you have loop devices. Facter 3 has a disks fact:
```console
$ facter --json disks
{
  "disks": {
    "vda": {
      "size": "50.00 GiB",
      "size_bytes": 53687091200,
      "vendor": "0x1af4"
    }
  }
}
```

Perhaps that's more reliable?

lzap replied:
Yes this is for historical reasons, discovery did not support facter3 facts. I think it should be fine to have this robust implementation for F3 and a simple fallback implementation for F2 just to show something.

domitea commented:
I hope that I've addressed all comments that are requested. I've raised issues in respective projects (Thanks @ares) and I've improve the parsing and added test with fixtures (Thanks @ekohl and @lzap ). However, from point of this PR, I've added only support for Facter version 3 and version 4. 

When I looked to the `puppet_fact_parser_test.rb`, I agree with @lzap that It should need rewrite test cases for working with all supported versions of Facter. I think that I have an idea how to do it but it's out of scope of PR. So maybe in future. Hovewer, since we trying to move Puppet out of core, it will be great if we're drop the support for Facter 2 (https://community.theforeman.org/t/the-road-to-making-puppet-optional/17983). So I think it would be great during this rewriting of test cases to drop support of Facter 2 also. That's the reason why I don't add in this fact support for Facter version 2.

ekohl commented:
Haven't looked through the changes, but wanted to comment on this:

> That's the reason why I don't add in this fact support for Facter version 2.

:+1: I think it's better to not provide the fact if there is no data than to make something up. Especially with Facter 2 on its way out.

ares commented:
The tests are failing

```
NoMethodError: undefined method `any?' for nil:NilClass
    app/services/puppet_fact_parser.rb:147:in `disks_total'
    app/models/host_facets/reported_data_facet.rb:11:in `populate_fields_from_facts'
    app/models/concerns/facets/base_host_extensions.rb:26:in `block in populate_facet_fields'
    app/models/concerns/facets/base_host_extensions.rb:25:in `each'
    app/models/concerns/facets/base_host_extensions.rb:25:in `populate_facet_fields'
    app/models/host/managed.rb:501:in `populate_fields_from_facts'
    app/models/concerns/puppet_host_extensions.rb:3:in `populate_fields_from_facts'
    app/services/host_fact_importer.rb:50:in `block (2 levels) in parse_facts'
    app/services/foreman/telemetry_helper.rb:27:in `telemetry_duration_histogram'
    app/services/host_fact_importer.rb:49:in `block in parse_facts'
    app/services/host_fact_importer.rb:90:in `block in skipping_orchestration'
    app/models/concerns/orchestration.rb:124:in `without_orchestration'
    app/services/host_fact_importer.rb:89:in `skipping_orchestration'
    app/services/host_fact_importer.rb:45:in `parse_facts'
    app/services/host_fact_importer.rb:34:in `import_facts'
    test/unit/host_fact_importer_test.rb:113:in `block in <class:HostFactImporterTest>' (NoMethodError)
/home/jenkins/workspace/test_develop_pr_core/database/postgresql/ruby/2.7/slave/fast/app/services/puppet_fact_parser.rb:147    
```

tbrisker commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,19 @@ def cores
+        disks_size += (disk_attrs['size_bytes'].to_f rescue 0)
+      end
+      # Turning disks_size to closest Mega for easier to read UI
+      (disks_size / 1024 / 1024).ceil if disks_size > 0
```
You can use
```suggestion
      disks_size / 1.megabyte
```
for easier readability. 
Also, there is no need to only do it for >0 since 0/1.megabyte is still 0.

tbrisker commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,19 @@ def cores
+
+    if disks.any?
+      disks.values.each do |disk_attrs|
+        disks_size += (disk_attrs['size_bytes'].to_f rescue 0)
```
`rescue`s in ruby are quite expensive, and fact parsing is a very performance-heavy area, so it would be better to handle differently. For example,
```suggestion
        disks_size += disk_attrs&.fetch('size_bytes', 0).to_f
```
Do we actually need this in number a float?

ekohl replied:
If they're bytes, it doesn't make sense to use a float since there is no unit smaller than a byte. A short and safe way to write it would be:
```ruby
disks_size = disks.sum { |disk| disk&.fetch('size_bytes', 0) }
```

tbrisker commented on db/migrate/20210317090500_add_disks_total_to_host_facets.rb:
```diff
@@ -0,0 +1,9 @@
+class AddDisksTotalToHostFacets < ActiveRecord::Migration[6.0]
+  def up
+    add_column :host_facets_reported_data_facets, :disks_total, :integer
```
What's the largest total disks that we can store in an integer? is that enough or do we need a `bigint`?

tbrisker commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,21 @@ def cores
+
+    return nil unless disks
+
+    if disks.any?
```
Do we still need the `.any?` here if we return early in the previous line?

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,21 @@ def cores
+    if disks.any?
+      disks.values.each do |disk_attrs|
+        disks_size += disk_attrs&.fetch('size_bytes', 0).to_i
+      end
```
I think this should also work:
```suggestion
    disk_size = disks.sum { |disk| disk&.fetch('size_bytes', 0).to_i }
```

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,21 @@ def cores
+        disks_size += disk_attrs&.fetch('size_bytes', 0).to_i
+      end
+      # Turning disks_size to closest Mega for easier to read UI
+      disks_size / 1.megabyte
```
I don't think this is a good API. A fact parser should not do anything for the UI. Other places might want a different value. If you have a helper `human_readable(bytes)` you need to convert it back to bytes. In every conversion there can be a loss. I'd argue it should return bytes and let another layer deal with the desired precision.

ekohl commented on test/unit/puppet_fact_parser_test.rb:
```diff
@@ -473,6 +485,14 @@ def structured_networking_facts
     read_json_fixture('facts/facts_structured_networking.json')['facts']
   end
 
+  def example_v3_facts
```
Just want to point out that [facterdb](https://github.com/camptocamp/facterdb) exists. Not asking for a change, but it may be good to know it's out there. Sadly, it looks like no Facter 4 facts are present right now.

ekohl commented on app/models/host_facets/reported_data_facet.rb:
```diff
@@ -8,6 +8,7 @@ def self.populate_fields_from_facts(host, parser, type, source_proxy)
         ram: parser.ram,
         sockets: parser.sockets,
         cores: parser.cores,
+        disks_total: parser.disks_total,
```
Shouldn't there be a `disks_total` in the [base FactParser](https://github.com/theforeman/foreman/blob/develop/app/services/fact_parser.rb)? Even if it returns nil, it'll at least prevent errors when a non-PuppetFactParser parser is used.

ekohl commented on app/services/puppet_fact_parser.rb:
```diff
@@ -140,6 +140,14 @@ def cores
+
+    return nil unless disks
+
+    disks.sum { |disk| disk&.fetch('size_bytes', 0).to_i }
```
If we're going minimal, isn't this the same?
```suggestion
    facts['disks']&.sum { |disk| disk&.fetch('size_bytes', 0).to_i }
```

ekohl reviewed (commented):
Technically it doesn't make a difference, but in your commit message you have `fixes` as lower case while most commits use `Fixes`. It makes it easier to ignore when going over `git log --oneline`. It is also custom to line wrap the commit message. See https://chris.beams.io/posts/git-commit/#wrap-72 for more details.

domitea commented:
I've made a improvements regarding to @ekohl comments. Thanks for the tip for facterdb! I'll probably use it for tests refactoring.

ekohl commented on app/services/fact_parser.rb:
```diff
@@ -132,6 +132,10 @@ def sockets
   def cores
   end
 
+  # summed total size of disks of host
```
```suggestion
  # The summed total size (in bytes) of all disks of a host or nil if unsupported
```

domitea replied:
Thanks! Comment has been updated.

ekohl reviewed (dismissed):
Small nit in the documentation that should make it easier to consume the API, but otherwise this looks good.

ekohl commented:
> It is also custom to line wrap the commit message. See https://chris.beams.io/posts/git-commit/#wrap-72 for more details.

Have you had a chance to look at this as well?

domitea commented:
@ekohl I've updated a commit message as you requested.

ekohl reviewed (approved):
Thanks!

lzap commented:
Thanks for test, I was hoping you would also add 2.X facter versions because apparently they are broken at the moment in 2.4:

https://community.theforeman.org/t/puppet-fact-parser-is-failing-with-string-does-not-have-dig-method/23138

But it is probably worth doing in a new PR because some changes will be necessary to fix bug(s).

ekohl commented:
There is no facter 2.x version because it doesn't have those facts. At best you can look at partitions but if you have loop devices the number will be incorrect.

> https://community.theforeman.org/t/puppet-fact-parser-is-failing-with-string-does-not-have-dig-method/23138

That particular issue was stringify facts set to true which breaks things. We will not ever support that.
